### PR TITLE
Implement LED testing mode

### LINKED ISSUE #235: LED testing mode
As mentioned on Discord. It would be great to have a LED testing mode that allows to freely toggle every LED on the board.
Background for me is testing the case for Glasgow and ensuring that the visibility of LEDs is good but probably also handy for testing functioning of the LEDs in general for manufacturing (which I think is @esden primary reason for wanting it).

### PR DESCRIPTION
**Pull request by whitequark**
Fixes #235.

### CODE REVIEW

**Review by esden — COMMENTED** [review]
This is really useful already. Is there any way to cycle through the LEDs? This would help us detect shorts between the LEDs,

File: software/glasgow/applet/internal/selftest/__init__.py [hunk 1]
```diff
@@ -54,6 +56,8 @@ class SelfTestApplet(GlasgowApplet, name="selftest"):
     Currently, shorts and opens on I/O lines can be detected.
 
     Test modes:
+        * leds: test indicator LED functionality
+          (Vio will be enabled)
```
Comment (attie) [inline software/glasgow/applet/internal/selftest/__init__.py:60]: `... enabled at 3.3v`

File: software/glasgow/applet/internal/selftest/__init__.py [hunk 2]
```diff
@@ -143,6 +147,13 @@ def decode_pins(bits):
         for mode in args.modes or self.__default_modes:
             self.logger.info("running self-test mode %s", mode)
 
+            if mode == "leds":
+                await device.test_leds(0b1111)
+                await device.write_register(self.addr_leds, 0b11111)
```
Comment (attie) [inline software/glasgow/applet/internal/selftest/__init__.py:152]: @esden - do you _need_ or positively _want_ the ability to set a pattern on the LEDs? (i.e: `0b01010` then `0b10101` to check for solder bridges)

The user LEDs are on nearby balls, and the FX2/ICE/ACT/ERR LEDs are on neighbouring pins.
Could be something like below, but it's a bit of extra work.

`glasgow run selftest leds --user 0x15 --stat 0xa`

**Comment by whitequark** [comment]
> Is there any way to cycle through the LEDs? This would help us detect shorts between the LEDs,

Yes the firmware in the test mode allows that. Should the cycling be manual or automatic on delay?

File: software/glasgow/applet/internal/selftest/__init__.py (hunk 2 above)
Comment (whitequark) [reply software/glasgow/applet/internal/selftest/__init__.py:152]: We don't have an option to set parameters on modes in the selftest applet, they'd have to be converted to subcommands and you can only have on then :(

Comment (attie) [reply software/glasgow/applet/internal/selftest/__init__.py:152]: > Should the cycling be manual or automatic on delay?

Alternatively, could maybe set a "walking ones" pattern on a 500ms timer?

Comment (attie) [reply software/glasgow/applet/internal/selftest/__init__.py:152]: > We don't have an option to set parameters on modes in the selftest applet

Yeah, that's the "extra work" I alluded to...

**Comment by attie** [comment]
I also noticed that it offers a `--port` argument, but ignores it... e.g: this enables Vio for both A and B.

```
$ glasgow run selftest --help
usage: glasgow run selftest [-h] [--port SPEC] [MODE ...]

Diagnose hardware faults.

Currently, shorts and opens on I/O lines can be detected.

Test modes:
  * leds: test indicator LED functionality (Vio will be enabled)
  * pins-int: detect shorts on traces between FPGA and I/O buffers (no requirements)
  * pins-ext: detect shorts and opens on traces between FPGA and I/O connector (all pins on all I/O connectors must be floating)
  * pins-pull: detects faults in pull resistor circuits (all pins on all I/O connectors must be floating)
  * pins-loop: detect faults anywhere in the I/O circuits (pins A0:A7 must be connected to B0:B7)
  * voltage: detect ADC, DAC or LDO faults (on all ports, Vsense and Vio pins must be connected)
  * loopback: detect faults in USB FIFO traces (no requirements)

options:
  -h, --help   show this help message and exit

build arguments:
  --port SPEC  bind the applet to port SPEC (default: AB)

run arguments:
  MODE         run self-test mode MODE (default: pins-int loopback)
$ glasgow run selftest --port A leds
I: g.device.hardware: device already has bitstream ID 117022fe3be96c747dd94733ba19f467
I: g.cli: running handler for applet 'selftest'
I: g.applet.internal.selftest: running self-test mode leds
W: g.applet.internal.selftest: power cycle the device to restore LED function
I: g.applet.internal.selftest: self-test: PASS
```

**Comment by whitequark** [comment]
Yeah the selftest applet has bitrotted greatly. If the LED testing mode is going to be gated on it being fixed I'd probably just not bother with it.

File: software/glasgow/applet/internal/selftest/__init__.py (hunk 2 above)
Comment (esden) [reply software/glasgow/applet/internal/selftest/__init__.py:152]: yes a fixed pattern every half a second would be fine I think. Similarely a sequence like like 0b00000,  0b00001,  0b00011,  0b00111,  0b01111,  0b11111,  0b11110,  0b11100,  0b11000,  0b10000, with a faster cadence of 10hz would work too.

Comment (whitequark) [reply software/glasgow/applet/internal/selftest/__init__.py:152]: Happy to implement the latter as the `glasgow test` mode.

Comment (esden) [reply software/glasgow/applet/internal/selftest/__init__.py:152]: Sounds good, I don't mind if this is not part of the selftest applet. I am running meta scripts to run the manufacturing test anyways.

Comment (attie) [reply software/glasgow/applet/internal/selftest/__init__.py:152]: Yes please, thanks @whitequark !

Comment (whitequark) [reply software/glasgow/applet/internal/selftest/__init__.py]: Done

**Comment by attie** [comment]
>> I also noticed that it offers a --port argument, but ignores it... e.g: this enables Vio for both A and B.
>
> Yeah the selftest applet has bitrotted greatly. If the LED testing mode is going to be gated on it being fixed I'll probably just not bother with it.

I wouldn't worry about this one - I just wanted to draw attention... far from blocking this PR.

File: software/glasgow/applet/internal/selftest/__init__.py (hunk 1 above)
Comment (whitequark) [reply software/glasgow/applet/internal/selftest/__init__.py:60]: This isn't always true at this point, you could sometimes get 0V by the time you stop

**Review by attie — APPROVED** [review]
Very pretty, thank you!

**Comment by esden** [comment]
Tested! Looks great. Thank you! :)

**Pull request merged**